File: services/games-api/app/services/stripe_service.py
"""
Stripe payment service for handling payments and webhooks
"""
import stripe
from typing import Dict, Any, Optional
from fastapi import HTTPException, status
from app.core.config import settings

# Initialize Stripe
stripe.api_key = settings.STRIPE_SECRET_KEY


class StripeService:
    """Service for handling Stripe payments and webhooks"""
    
    def __init__(self):
        self.webhook_secret = None  # Would be set from environment in production
    
    async def create_payment_intent(
        self, 
        amount: int, 
        currency: str = "usd",
        metadata: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """
        Create a Stripe Payment Intent
        
        Args:
            amount: Amount in smallest currency unit (e.g., cents for USD)
            currency: Currency code (default: "usd")
            metadata: Additional metadata for the payment
            
        Returns:
            Stripe Payment Intent object
        """
        try:
            intent = stripe.PaymentIntent.create(
                amount=amount,
                currency=currency,
                metadata=metadata or {},
                automatic_payment_methods={
                    'enabled': True,
                },
            )
            return {
                'client_secret': intent.client_secret,
                'id': intent.id,
                'status': intent.status
            }
        except stripe.error.StripeError as e:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Stripe error: {str(e)}"
            )
    
    async def create_connect_account(self, user_data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Create a Stripe Connect account for a user
        
        Args:
            user_data: User information for account creation
            
        Returns:
            Stripe Account object
        """
        try:
            account = stripe.Account.create(
                type="express",
                country="US",
                email=user_data.get('email'),
                capabilities={
                    "card_payments": {"requested": True},
                    "transfers": {"requested": True},
                },
                business_type="individual",
                individual={
                    "first_name": user_data.get('first_name', ''),
                    "last_name": user_data.get('last_name', ''),
                    "email": user_data.get('email'),
                },
            )
            return {
                'id': account.id,
                'details_submitted': account.details_submitted
            }
        except stripe.error.StripeError as e:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Stripe Connect error: {str(e)}"
            )
    
    async def create_account_link(self, account_id: str) -> Dict[str, Any]:
        """
        Create an account link for Stripe Connect onboarding
        
        Args:
            account_id: Stripe Connect account ID
            
        Returns:
            Account link URL
        """
        try:
            link = stripe.AccountLink.create(
                account=account_id,
                refresh_url=f"{settings.FRONTEND_URL}/stripe/onboarding/refresh",
                return_url=f"{settings.FRONTEND_URL}/stripe/onboarding/success",
                type="account_onboarding",
            )
            return {'url': link.url}
        except stripe.error.StripeError as e:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Stripe account link error: {str(e)}"
            )
    
    async def verify_webhook_signature(self, payload: bytes, sig_header: str) -> bool:
        """
        Verify Stripe webhook signature
        
        Args:
            payload: Raw request payload
            sig_header: Stripe signature header
            
        Returns:
            Boolean indicating if signature is valid
        """
        try:
            event = stripe.Webhook.construct_event(
                payload, sig_header, self.webhook_secret
            )
            return True
        except stripe.error.SignatureVerificationError:
            return False
        except Exception:
            return False
    
    async def handle_payment_success(self, payment_intent_id: str) -> Dict[str, Any]:
        """
        Handle successful payment webhook
        
        Args:
            payment_intent_id: Stripe Payment Intent ID
            
        Returns:
            Payment information
        """
        try:
            payment_intent = stripe.PaymentIntent.retrieve(payment_intent_id)
            
            if payment_intent.status == 'succeeded':
                return {
                    'id': payment_intent.id,
                    'amount': payment_intent.amount,
                    'currency': payment_intent.currency,
                    'metadata': payment_intent.metadata,
                    'status': payment_intent.status
                }
            else:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail=f"Payment not succeeded: {payment_intent.status}"
                )
        except stripe.error.StripeError as e:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Stripe error: {str(e)}"
            )
    
    async def create_payout(self, account_id: str, amount: int, currency: str = "usd") -> Dict[str, Any]:
        """
        Create a payout to a connected account
        
        Args:
            account_id: Stripe Connect account ID
            amount: Amount in smallest currency unit
            currency: Currency code
            
        Returns:
            Payout information
        """
        try:
            payout = stripe.Payout.create(
                amount=amount,
                currency=currency,
                destination=account_id,
            )
            return {
                'id': payout.id,
                'amount': payout.amount,
                'currency': payout.currency,
                'status': payout.status
            }
        except stripe.error.StripeError as e:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Stripe payout error: {str(e)}"
            )


# Global Stripe service instance
stripe_service = StripeService()

File: services/booking/src/payments/stripe/stripe.service.ts

import { Injectable, Logger } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import Stripe from 'stripe';
import { PaymentRepository } from '../payment.repository';

@Injectable()
export class StripeService {
  private stripe: Stripe;
  private readonly logger = new Logger(StripeService.name);

  constructor(
    private configService: ConfigService,
    private paymentRepository: PaymentRepository
  ) {
    const stripeSecretKey = this.configService.get('STRIPE_SECRET_KEY');
    this.logger.log(`STRIPE_SECRET_KEY: ${stripeSecretKey ? 'SET' : 'NOT SET'}`);
    
    this.stripe = new Stripe(stripeSecretKey || '', {
      apiVersion: '2023-10-16',
    });
  }

  async createConnectAccount(email: string, userId: string): Promise<any> {
    try {
      this.logger.log(`Creating Stripe Connect account for user: ${userId}, email: ${email}`);
      
      if (this.configService.get('STRIPE_SECRET_KEY')?.includes('test')) {
        this.logger.debug('Using test mode - returning mock account');
        return {
          id: `acct_test_${Date.now()}`,
          object: 'account',
          email,
          metadata: { userId },
          type: 'express',
          created: Math.floor(Date.now() / 1000),
          capabilities: {
            card_payments: { status: 'pending', requested: true },
            transfers: { status: 'pending', requested: true },
          },
        };
      }

      const account = await this.stripe.accounts.create({
        type: 'express',
        email,
        metadata: {
          userId,
        },
        capabilities: {
          card_payments: { requested: true },
          transfers: { requested: true },
        },
      });

      this.logger.log(`Successfully created Stripe Connect account: ${account.id}`);
      return account;
    } catch (error) {
      this.logger.error(`Failed to create Stripe Connect account: ${error.message}`, error.stack);
      throw new Error(`Failed to create Stripe Connect account: ${error.message}`);
    }
  }

  async createAccountLink(accountId: string, refreshUrl: string, returnUrl: string): Promise<any> {
    try {
      this.logger.log(`Creating account link for account: ${accountId}`);
      
      if (this.configService.get('STRIPE_SECRET_KEY')?.includes('test')) {
        this.logger.debug('Using test mode - returning mock account link');
        return {
          object: 'account_link',
          created: Math.floor(Date.now() / 1000),
          expires_at: Math.floor(Date.now() / 1000) + 3600,
          url: `https://connect.stripe.com/setup/test/${accountId}`,
        };
      }

      const accountLink = await this.stripe.accountLinks.create({
        account: accountId,
        refresh_url: refreshUrl,
        return_url: returnUrl,
        type: 'account_onboarding',
      });

      this.logger.log(`Successfully created account link for account: ${accountId}`);
      return accountLink;
    } catch (error) {
      this.logger.error(`Failed to create account link: ${error.message}`, error.stack);
      throw new Error(`Failed to create account link: ${error.message}`);
    }
  }

  async getAccountStatus(accountId: string): Promise<any> {
    try {
      this.logger.log(`Getting account status for: ${accountId}`);
      
      if (this.configService.get('STRIPE_SECRET_KEY')?.includes('test')) {
        this.logger.debug('Using test mode - returning mock account status');
        return {
          id: accountId,
          charges_enabled: false,
          payouts_enabled: false,
          requirements: {
            currently_due: [],
            eventually_due: [],
            past_due: [],
            pending_verification: [],
          },
          details_submitted: false,
        };
      }

      const account = await this.stripe.accounts.retrieve(accountId);
      
      this.logger.log(`Retrieved account status for: ${accountId} - charges: ${account.charges_enabled}, payouts: ${account.payouts_enabled}`);
      return {
        id: account.id,
        charges_enabled: account.charges_enabled,
        payouts_enabled: account.payouts_enabled,
        requirements: account.requirements,
        details_submitted: account.details_submitted,
      };
    } catch (error) {
      this.logger.error(`Failed to get account status: ${error.message}`, error.stack);
      throw new Error(`Failed to get account status: ${error.message}`);
    }
  }

  async createPaymentIntent(
    amount: number,
    currency: string,
    connectedAccountId: string,
    applicationFeeAmount: number,
    metadata: Record<string, string> = {}
  ): Promise<any> {
    try {
      this.logger.log(`Creating payment intent for connected account: ${connectedAccountId}, amount: ${amount}, fee: ${applicationFeeAmount}`);
      
      if (this.configService.get('STRIPE_SECRET_KEY')?.includes('test')) {
        this.logger.debug('Using test mode - returning mock payment intent');
        return {
          id: `pi_test_${Date.now()}`,
          object: 'payment_intent',
          amount,
          currency,
          status: 'requires_payment_method',
          client_secret: `pi_test_${Date.now()}_secret`,
          application_fee_amount: applicationFeeAmount,
          metadata,
        };
      }

      const paymentIntent = await this.stripe.paymentIntents.create({
        amount,
        currency,
        application_fee_amount: applicationFeeAmount,
        transfer_data: {
          destination: connectedAccountId,
        },
        metadata,
        automatic_payment_methods: {
          enabled: true,
        },
      });

      this.logger.log(`Successfully created payment intent: ${paymentIntent.id}`);
      return paymentIntent;
    } catch (error) {
      this.logger.error(`Failed to create payment intent: ${error.message}`, error.stack);
      throw new Error(`Failed to create payment intent: ${error.message}`);
    }
  }

  async createRefund(
    paymentIntentId: string,
    amount?: number,
    reason?: string
  ): Promise<any> {
    try {
      this.logger.log(`Creating refund for payment intent: ${paymentIntentId}, amount: ${amount}`);
      
      if (this.configService.get('STRIPE_SECRET_KEY')?.includes('test')) {
        this.logger.debug('Using test mode - returning mock refund');
        return {
          id: `re_test_${Date.now()}`,
          object: 'refund',
          amount: amount || 1000,
          currency: 'usd',
          status: 'succeeded',
          payment_intent: paymentIntentId,
        };
      }

      const refund = await this.stripe.refunds.create({
        payment_intent: paymentIntentId,
        amount,
        reason: reason as any, // Type assertion for Stripe enum
      });

      this.logger.log(`Successfully created refund: ${refund.id}`);
      return refund;
    } catch (error) {
      this.logger.error(`Failed to create refund: ${error.message}`, error.stack);
      throw new Error(`Failed to create refund: ${error.message}`);
    }
  }

  constructEvent(payload: Buffer, signature: string, secret: string): any {
    try {
      this.logger.log(`Constructing Stripe event with signature: ${signature.substring(0, 20)}...`);
      const event = this.stripe.webhooks.constructEvent(payload, signature, secret);
      this.logger.log(`Successfully constructed Stripe event: ${event.type} (${event.id})`);
      return event;
    } catch (error) {
      this.logger.error(`Failed to construct Stripe event: ${error.message}`, error.stack);
      throw new Error(`Failed to construct Stripe event: ${error.message}`);
    }
  }
}
